File: oxidize-server/src/schema.rs
//! OpenAI-compatible request/response schemas.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Deserialize)]
pub struct ChatCompletionRequest {
    pub model: String,
    pub messages: Vec<ChatMessageInput>,
    #[serde(default)]
    pub response_format: Option<ResponseFormat>,
    #[serde(default)]
    pub guided_json: Option<Value>,
    #[serde(default)]
    pub json_schema: Option<Value>,
    #[serde(default)]
    pub guided_regex: Option<String>,
    #[serde(default)]
    pub guided_choice: Option<Vec<String>>,
    #[serde(default)]
    pub stream: bool,
    #[serde(default)]
    pub max_tokens: Option<usize>,
    #[serde(default)]
    pub max_completion_tokens: Option<usize>,
    #[serde(default)]
    pub temperature: Option<f32>,
    #[serde(default)]
    pub top_p: Option<f32>,
    #[serde(default)]
    pub top_k: Option<usize>,
    #[serde(default)]
    pub min_p: Option<f32>,
    #[serde(default)]
    pub typical_p: Option<f32>,
    #[serde(default)]
    pub tail_free_z: Option<f32>,
    #[serde(default)]
    pub stop: Option<StopSequences>,
    #[serde(default)]
    pub seed: Option<u64>,
    #[serde(default)]
    pub n: Option<usize>,
    #[serde(default)]
    pub best_of: Option<usize>,
}

#[derive(Debug, Deserialize)]
pub struct ChatMessageInput {
    pub role: String,
    pub content: String,
    #[serde(default)]
    pub images: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
pub struct CompletionRequest {
    pub model: String,
    pub prompt: String,
    #[serde(default)]
    pub response_format: Option<ResponseFormat>,
    #[serde(default)]
    pub guided_json: Option<Value>,
    #[serde(default)]
    pub json_schema: Option<Value>,
    #[serde(default)]
    pub guided_regex: Option<String>,
    #[serde(default)]
    pub guided_choice: Option<Vec<String>>,
    #[serde(default)]
    pub stream: bool,
    #[serde(default)]
    pub max_tokens: Option<usize>,
    #[serde(default)]
    pub temperature: Option<f32>,
    #[serde(default)]
    pub top_p: Option<f32>,
    #[serde(default)]
    pub top_k: Option<usize>,
    #[serde(default)]
    pub min_p: Option<f32>,
    #[serde(default)]
    pub typical_p: Option<f32>,
    #[serde(default)]
    pub tail_free_z: Option<f32>,
    #[serde(default)]
    pub stop: Option<StopSequences>,
    #[serde(default)]
    pub seed: Option<u64>,
    #[serde(default)]
    pub echo: bool,
    #[serde(default)]
    pub n: Option<usize>,
    #[serde(default)]
    pub best_of: Option<usize>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum StopSequences {
    One(String),
    Many(Vec<String>),
}

impl StopSequences {
    pub fn into_vec(self) -> Vec<String> {
        match self {
            Self::One(value) => vec![value],
            Self::Many(values) => values,
        }
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ResponseFormat {
    Text,
    JsonObject,
    JsonSchema { json_schema: Value },
}

impl ResponseFormat {
    pub fn output_text(&self) -> &'static str {
        match self {
            Self::Text => "",
            Self::JsonObject => "{}",
            Self::JsonSchema { json_schema } => {
                let _ = json_schema;
                "{}"
            }
        }
    }
}

#[derive(Debug, Deserialize)]
pub struct EmbeddingsRequest {
    pub model: String,
    pub input: Value,
}

#[derive(Debug, Serialize)]
pub struct ModelsResponse {
    pub object: &'static str,
    pub data: Vec<ModelData>,
}

#[derive(Debug, Serialize)]
pub struct ModelData {
    pub id: String,
    pub object: &'static str,
    pub owned_by: &'static str,
}

File: golang/internal/api/schema.go
package api

import (
	"encoding/json"
	"fmt"
)

type ResponseFormatType string

const (
	ResponseFormatText       ResponseFormatType = "text"
	ResponseFormatJSONObject ResponseFormatType = "json_object"
	ResponseFormatJSONSchema ResponseFormatType = "json_schema"
)

type ChatCompletionRequest struct {
	Model               string          `json:"model"`
	Messages            []ChatMessage   `json:"messages"`
	ResponseFormat      *ResponseFormat `json:"response_format,omitempty"`
	GuidedJSON          json.RawMessage `json:"guided_json,omitempty"`
	JSONSchema          json.RawMessage `json:"json_schema,omitempty"`
	GuidedRegex         string          `json:"guided_regex,omitempty"`
	GuidedChoice        []string        `json:"guided_choice,omitempty"`
	Stream              bool            `json:"stream,omitempty"`
	MaxTokens           *int            `json:"max_tokens,omitempty"`
	MaxCompletionTokens *int            `json:"max_completion_tokens,omitempty"`
	Temperature         *float64        `json:"temperature,omitempty"`
	TopP                *float64        `json:"top_p,omitempty"`
	TopK                *int            `json:"top_k,omitempty"`
	MinP                *float64        `json:"min_p,omitempty"`
	TypicalP            *float64        `json:"typical_p,omitempty"`
	TailFreeZ           *float64        `json:"tail_free_z,omitempty"`
	Stop                StopSequences   `json:"stop,omitempty"`
	Seed                *uint64         `json:"seed,omitempty"`
	N                   *int            `json:"n,omitempty"`
	BestOf              *int            `json:"best_of,omitempty"`
}

type ChatMessage struct {
	Role    string   `json:"role"`
	Content string   `json:"content"`
	Images  []string `json:"images,omitempty"`
}

type CompletionRequest struct {
	Model          string          `json:"model"`
	Prompt         string          `json:"prompt"`
	ResponseFormat *ResponseFormat `json:"response_format,omitempty"`
	GuidedJSON     json.RawMessage `json:"guided_json,omitempty"`
	JSONSchema     json.RawMessage `json:"json_schema,omitempty"`
	GuidedRegex    string          `json:"guided_regex,omitempty"`
	GuidedChoice   []string        `json:"guided_choice,omitempty"`
	Stream         bool            `json:"stream,omitempty"`
	MaxTokens      *int            `json:"max_tokens,omitempty"`
	Temperature    *float64        `json:"temperature,omitempty"`
	TopP           *float64        `json:"top_p,omitempty"`
	TopK           *int            `json:"top_k,omitempty"`
	MinP           *float64        `json:"min_p,omitempty"`
	TypicalP       *float64        `json:"typical_p,omitempty"`
	TailFreeZ      *float64        `json:"tail_free_z,omitempty"`
	Stop           StopSequences   `json:"stop,omitempty"`
	Seed           *uint64         `json:"seed,omitempty"`
	Echo           bool            `json:"echo,omitempty"`
	N              *int            `json:"n,omitempty"`
	BestOf         *int            `json:"best_of,omitempty"`
}

type EmbeddingsRequest struct {
	Model string          `json:"model"`
	Input json.RawMessage `json:"input"`
}

type ResponseFormat struct {
	Type       ResponseFormatType `json:"type"`
	JSONSchema json.RawMessage    `json:"json_schema,omitempty"`
}

func (f ResponseFormat) OutputText() string {
	switch f.Type {
	case ResponseFormatJSONObject, ResponseFormatJSONSchema:
		return "{}"
	default:
		return ""
	}
}

type StopSequences struct {
	values []string
}

func (s *StopSequences) UnmarshalJSON(data []byte) error {
	var single string
	if err := json.Unmarshal(data, &single); err == nil {
		s.values = []string{single}
		return nil
	}
	var many []string
	if err := json.Unmarshal(data, &many); err == nil {
		s.values = append([]string(nil), many...)
		return nil
	}
	return fmt.Errorf("decode stop sequences")
}

func (s StopSequences) Values() []string {
	return append([]string(nil), s.values...)
}
